Documentation/design-docs/framework-version-resolution.md

```diff
@@ -0,0 +1,290 @@
+      }
+    ]
+  }
+}
+```
+
+3. Environment variable `DOTNET_ROLL_FORWARD` - For example:
+```
+set DOTNET_ROLL_FORWARD=LatestMajor
+```
+
+4. Command line argument `--roll-forward` - For example:
+```
+dotnet --roll-forward LatestMinor app.dll
+```
+
+For the framework references specified by the application itself:  
+The order above defines precedence. Later scopes take precedence over earlier ones. So command line wins over everything, environment variable wins over `.runtimeconfig.json` and per-framework setting wins over config-wide setting.
+
+For dependent frameworks:
+Each dependent framework (for example FX1) is loaded with an effective roll-forward value (the one used to resolve that framework). If FX1 has additional framework references those will be loaded using this same roll-forward value as the default, that is unless FX1 specifies different value in its `.runtimeconfig.json` or there's a higher precedence override (environment variable `DOTNET_ROLL_FORWARD` or command line argument `--roll-forward`).
+
+
+## Pre-release versions
+Pre-release version is a version which has a pre-release part, for example `3.0.0-preview4-27415-15`. Everything after the `-` (dash) character is a pre-release identifier. Per [semantic versioning rules](https://semver.org/) pre-release versions are ordered before the same version without any pre-release part. So `3.0.0-preview4-27415-15` comes before `3.0.0`.
+
+### Old behavior
+Before .NET Core 3.0 (so 2.2 and older) the roll-forward behavior for pre-release versions ignored any of the roll-forward related settings, that is both `rollForwardOnNoCandidateFx` as well as `applyPatches`.
+The behavior was that:
+* Release version will never roll forward to pre-release version. So for example `2.1.0` will not roll forward to `3.0.0-preview4-27415-15` even if `Major` roll-forward is allowed (same for all the other settings of `rollForwardOnNoCandidateFx` as well as `applyPatches`).
+* Pre-release version will never roll forward to release version. So for example `3.0.0-preview4-27415-15` will not roll forward to `3.0.0`. Also pre-release will only roll forward to the same `major.minor.patch`. So for example `3.0.0-preview4-27415-15` will not roll forward to `3.0.1-preview1-29000-0`.
+
+### New behavior
+The newly proposed behavior is to treat pre-release versions no different from release versions. This would mean that pre-release can roll forward to release but also that release can roll forward to pre-release. For the most part this improves the previous behavior (except when `Disable` is active in which case no roll-forward will happen):
+* `3.0.0-preview4-27415-15` will roll forward to `3.0.0`.
+* `3.0.0-preview4-27415-15` will roll forward to `3.0.1-preview1-29000-0`.
+
+But it also introduces cases which are somewhat controversial:
+* On `Minor` (the default) `3.0.0` would roll forward to `3.0.1-preview1-29000-0` even if `3.0.1` is available on the machine. This behavior is typically not expected.
```

Comment:
Consider the real case of a build called X.Y.N-* where N > 0: 

1. These are bits that are about to go out as a patch.
2. We never advertise these builds to the public, we only use them to validate the upcoming patch.

Behaviors of these builds that are different from final X.Y.N work against (2).

Reply:
```
treat pre-release versions no different from release versions
```

Seems like a bad policy which violates SemVer2 rules.  SemV2 would allow the prerelease to satisfy the release version, but would not be equal to the release version, but actually less than it.  


```
On `Minor` (the default) `3.0.0` would roll forward to `3.0.1-preview1-29000-0` even if `3.0.1` is
available on the machine.
```

Seems like we should just use SemVer2 rules for prerelease. For all intents and purposes `3.0.1` is the same as `3.0.1-preview1-29000-0`, except when compared against each other, in which case the release should be preferred.

`3.0.1-preview1-29000-0` should roll forward to its release version `3.0.1`.  This is consistent with latest patch ordering.

Reply:
Yes, I don't think anyone meant to suggest `3.0.1-preview > 3.0.1`, maybe just informal language not being clear somewhere, but I can't see where you're quoting from.

Reply:
I had a but in the spec before... Steve pointed it out correctly.

Reply:
Oops, I see now.